src/theory/booleans/proof_checker.cpp

```diff
@@ -190,21 +191,147 @@ Node BoolProofRuleChecker::checkInternal(PfRule id,
   if (id == PfRule::CHAIN_RESOLUTION)
   {
     Assert(children.size() > 1);
-    Assert(args.size() == children.size() - 1);
+    Assert(args.size() == 2 * (children.size() - 1));
     Trace("bool-pfcheck") << "chain_res:\n" << push;
+    NodeManager* nm = NodeManager::currentNM();
+    Node trueNode = nm->mkConst(true);
+    Node falseNode = nm->mkConst(false);
     std::vector<Node> clauseNodes;
-    for (unsigned i = 0, childrenSize = children.size(); i < childrenSize; ++i)
+    // literals to be removed from the virtual lhs clause of the resolution
+    std::unordered_map<Node, unsigned, NodeHashFunction> lhsElim;
+    for (std::size_t i = 0, argsSize = args.size(); i < argsSize; i = i + 2)
+    {
+      // whether pivot should occur as is or negated depends on the id of
+      // each step in the chain
+      if (args[i] == trueNode)
+      {
+        lhsElim[args[i + 1]]++;
+      }
+      else
+      {
+        Assert(args[i] == falseNode);
+        lhsElim[args[i + 1].notNode()]++;
+      }
+    }
+    if (Trace.isOn("bool-pfcheck"))
+    {
+      Trace("bool-pfcheck")
+          << "Original elimination multiset for lhs clause:\n";
+      for (const auto& pair : lhsElim)
+      {
+        Trace("bool-pfcheck")
+            << "\t- " << pair.first << " {" << pair.second << "}\n";
+      }
+    }
+    for (std::size_t i = 0, childrenSize = children.size(); i < childrenSize;
+         ++i)
+    {
+      // literal to be removed from rhs clause. They will be negated
+      Node rhsElim = Node::null();
+      if (Trace.isOn("bool-pfcheck"))
+      {
+        Trace("bool-pfcheck") << i << ": current lhsElim:\n";
+        for (const auto& pair : lhsElim)
+        {
+          Trace("bool-pfcheck")
+              << "\t- " << pair.first << " {" << pair.second << "}\n";
+        }
+      }
+      if (i > 0)
+      {
+        std::size_t index = 2 * (i - 1);
+        rhsElim = args[index] == trueNode ? args[index + 1].notNode()
+                                          : args[index + 1];
+        Trace("bool-pfcheck") << i << ": rhs elim: " << rhsElim << "\n";
+      }
+      // only add to conclusion nodes that are not in elimination set. First get
+      // the nodes.
+      //
+      // Since unit clauses can also be OR nodes, we rely on the invariant that
+      // non-unit clauses will not occur themselves in their elimination sets.
+      // If they do then they must be unit.
```

Comment:
First: I always feel "unit clause" is reserved to SAT solvers, and we should rather use "singleton clause"...

Then I'm always confused by what the code below does. Suggestion for a different (possibly better) comment:

```suggestion
      // Since a Node can not hold an OR with a single child we need to take of such singleton clauses.
      // If the child is not an OR, it is a singleton clause and we only take the child itself.
      // Otherwise, it can only be a singleton clause if the child itself is used as a resolution literal, that is if the child is in lhsElim or rhsElim (and then the negation of the child is in lhsElim).
```

Reply:
I'll change accordingly, thanks!